### PR TITLE
Fix Error Message for RunN Workflow IDs

### PR DESCRIPTION
**Pull request by datamel**
These changes close https://github.com/cylc/cylc-flow/issues/4766

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Appropriate tests are included (unit and/or functional).
- [x] No change log entry required (Error change).
- [x] No documentation update required.

### CODE REVIEW

File: cylc/flow/workflow_files.py [hunk 1]
```diff
@@ -1246,14 +1247,22 @@ def infer_latest_run(
...
     try:
         reg = path.relative_to(cylc_run_dir)
     except ValueError:
         raise ValueError(f"{path} is not in the cylc-run directory")
+    if not path.exists():
+        raise UserInputError(f'Path does not exist: {path}')
     if path.name == WorkflowFiles.RUN_N:
+        LOG.warning(
+            f"The use of {WorkflowFiles.RUN_N} in the Workflow ID is not"
+            f" necessary."
```
Comment (hjoliver) [inline cylc/flow/workflow_files.py:1264]: Not a biggie, but this ought to be an infrequently-seen warning, so I think we could afford to be more explicit, and take the opportunity to explain what the `runN` symlink is actually for:
```suggestion
            f"Explicit use of {WorkflowFiles.RUN_N} in the Workflow ID is not"
            f" necessary. It is used automatically to select the latest run number."
```

File: tests/unit/test_id_cli.py [hunk 2]
```diff
@@ -56,7 +57,9 @@ def abc_src_dir(tmp_path_factory):
 )
 async def test_parse_ids_workflows(ids_in, ids_out):
     """It should parse workflows & tasks."""
-    workflows, _ = await parse_ids_async(*ids_in, constraint='workflows')
+    with patch.object(Path, 'exists') as mock_exists:
```
Comment (wxtim) [inline tests/unit/test_id_cli.py:60]: There are a lot of very similar context managers here - I wonder if you could turn all these into a fixture?

File: tests/unit/test_workflow_files.py [hunk 3]
```diff
@@ -301,13 +301,32 @@ def test_infer_latest_run(
         assert infer_latest_run(path) == expected
 
 
+def test_infer_latest_run_raise_warning_for_runN(caplog, tmp_run_dir):
+    caplog.set_level(logging.WARNING, 'log')
+    cylc_run_dir: Path = tmp_run_dir()
```
Comment (wxtim) [inline tests/unit/test_workflow_files.py:306]: I don't think that this assignment is necessary either - `temp_run_dir` has the default fixture scope of function so you can dispense with this line if you wish, just using `tmp_run_dir()` each time you need it.

File: tests/unit/test_workflow_files.py [hunk 4]
```diff
@@ -301,13 +301,32 @@ def test_infer_latest_run(
...
+    run_dir = cylc_run_dir / 'run1'
+    run_dir.mkdir()
+    runN_path = cylc_run_dir / 'runN'
+    runN_path.symlink_to('run1')
+    infer_latest_run(runN_path)
+    assert ("The use of runN in the Workflow ID is not necessary."
```
Comment (wxtim) [inline tests/unit/test_workflow_files.py:312]: Consider using a regex or similar to check that the essentials of this message are there, without making it very fragile in terms of exact words used?

**Review by wxtim — APPROVED** [review]
Sorry, I went a bit pedantic on your test. It works, so feel free to implement or not my suggestions.

I also agree with Hilary's comment on verbosity of warning.

Approved.

**Review by hjoliver — APPROVED** [review]

**Pull request merged**